src/yb/tablet/tablet_bootstrap.cc

```diff
@@ -316,13 +316,35 @@ Status ReplayState::UpdateSplitOpId(const ReplicateMsg& msg, const TabletId& tab
 }
 
 void ReplayState::AddEntriesToStrings(const OpIndexToEntryMap& entries,
-                                      std::vector<std::string>* strings) const {
+                                      std::vector<std::string>* strings,
+                                      int limit) const {
+  int n = entries.size();
```

Comment:
Let's create a utility function (as a lambda -- only used in the same function).

```c++
auto append_metadata_for_range = [&](size_t start_index, size_t num_records) {
  for (auto i = start_index; i < start_index + num_records; ++i) {
    const auto& replicate = entries[i];
    strings->push_back(Format(
        "    [$0] op_id: $1 hybrid_time: $2 op_type: $3 committed_op_id: $4",
        i + 1,
        OpId::FromPB(replicate.id()),
        replicate.hybrid_time(),
        replicate.op_type(),
        OpId::FromPB(replicate.committed_op_id())));
  }
};
```

Note that I've replaced `id` -> `op_id` as this is a more common notation for Raft OpIds in our codebase. Also, note the use of `OpId::FromPB` which will convert the OpIdPB to the non-protobuf OpId struct that then gets automatically formatted as "<term>.<index>".

Then, I think the logic that follows (with "first", "second", "truncated", etc.) is still too complicated. Let's replace it with the following:

```c++
const auto n = entries.size();
if (n <= 2 * half_limit) {
  append_metadata_for_range(0, n);
  return;
}
append_metadata_for_range(0, half_limit);
strings->push_back(Format("($0 lines skipped)", n - 2 * half_limit));
append_metadata_for_range(n - half_limit, half_limit);
```

Reply:
w.r.t. entries[I], I get the following compilation error:
```
src/yb/tablet/tablet_bootstrap.cc:340:22: error: non-const lvalue reference to type 'log::LogEntryPB' cannot bind to a temporary of type 'std::__1::unique_ptr<yb::log::LogEntryPB, std::__1::default_delete<yb::log::LogEntryPB> >::pointer' (aka 'yb::log::LogEntryPB *')
    log::LogEntryPB& entry = map_entry.second.entry.get();

------
src/yb/tablet/tablet_bootstrap.cc:324:38: error: no viable overloaded operator[] for type 'const yb::tablet::OpIndexToEntryMap' (aka 'const map<long long, yb::tablet::Entry>')
      const auto& replicate = entries[i];
                              ~~~~~~~^~
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/map:1126:18: note: candidate function not viable: 'this' argument has type 'const yb::tablet::OpIndexToEntryMap' (aka 'const map<long long, yb::tablet::Entry>'), but method is not marked const
    mapped_type& operator[](const key_type& __k);
                 ^
/Library/Developer/CommandLineTools/usr/bin/../include/c++/v1/map:1128:18: note: candidate function not viable: 'this' argument has type 'const yb::tablet::OpIndexToEntryMap' (aka 'const map<long long, yb::tablet::Entry>'), but method is not marked const
    mapped_type& operator[](key_type&& __k);
```

Reply:
In order to use the [] syntax to access the map, I had to change the method declaration for AddEntriesToStrings (dropping the const):
```
void ReplayState::AddEntriesToStrings(OpIndexToEntryMap& entries,
                                      std::vector<std::string>* strings,
```
Now I have:
```
      const auto& replicate = entries[i].entry->replicate();
```
But there is still one compilation error:
```
src/yb/tablet/tablet_bootstrap.cc:365:25: error: binding value of type 'const map<...>' to reference to type 'map<...>' drops 'const' qualifier
    AddEntriesToStrings(pending_replicates, strings, (limit-cnt)/2);
                        ^~~~~~~~~~~~~~~~~~
src/yb/tablet/tablet_bootstrap.cc:318:58: note: passing argument to parameter 'entries' here
void ReplayState::AddEntriesToStrings(OpIndexToEntryMap& entries,
                                                         ^
```

Reply:
OK, sorry, it's a map. Of course `[]` won't work here, and changing it to non-const is not an option. Then I would use something like this, along the lines of your solution, but with fewer cases to reason about (with min/max, etc.)
```c++
const auto n = entries.size();
const bool overflow = n > 2 * half_limit;
size_t index = 0;
for (const auto& entry : entries) {
  if (!overflow || (index < half_limit || index >= n - half_limit)) {
    const auto& replicate = entry.second.entry->replicate();
    strings->push_back(Format(
        "    [$0] op_id: $1 hybrid_time: $2 op_type: $3 committed_op_id: $4",
        index + 1,
        OpId::FromPB(replicate.id()),
        replicate.hybrid_time(),
        replicate.op_type(),
        OpId::FromPB(replicate.committed_op_id())));
  };
  if (overflow && index == half_limit - 1) {
    strings->push_back(Format("($0 lines skipped)", n - 2 * half_limit));
  }
  index++;
}
```